feat!: when filling dask_histgram.boost.Histograms delay creation of task graph, use multi-fill

lgray opened the pull request:
Closes #126  (supercedes, github was stuck)

Refer to that PR for justification.

This PR implements:
- reverts the current tree reduce pattern for filling dask-boost-histograms
- delayed generation of dask-boost-histogram taskgraphs
- multi-fills where in the case of dask-awkward array inputs all fills to a histogram are grouped into a single task
  - this is implemented as a tree reduce

lgray commented:
@martindurant OK there was a little bit of cleanup to do - it's good to go now.

lgray commented:
`split_every` -> `split-every` in `histogram.yaml` to be congruent with dask-awkward's formatting choices.

martindurant commented on src/dask_histogram/boost.py:
```diff
@@ -242,12 +190,73 @@ def dask_name(self) -> str:
+        )
+
+        if is_dask_awkward_like(dask_data[0]):
+
```
This being such a big if, comments describing the two arm definitely helpful.
The first case if where (multiple) dak arrays have been used to fill the hist. How come we only look at the very first dask collection?

lgray replied:
this is removed now and dealt with more appropriately

martindurant commented on src/dask_histogram/boost.py:
```diff
@@ -242,12 +190,73 @@ def dask_name(self) -> str:
+            for datum in (
+                self._staged[0]["args"] + tuple(self._staged[0]["kwargs"].values())
+            )
+            if is_dask_collection(datum)
```
What becomes of concrete, non-dask values?

lgray replied:
this is now dealt with more appropriately.

martindurant commented on src/dask_histogram/boost.py:
```diff
@@ -242,12 +190,73 @@ def dask_name(self) -> str:
+            output_hist = _reduction(fills, split_every)
+        else:
+
+            first_fill = self._staged.pop()
```
This is equivalent to the previous behaviour, right? So in some future case, we may still want to know what to do with, for example, dask.array fills.

lgray replied:
this has been entirely changed now - all input variants use the multifill strategy now.

martindurant commented on src/dask_histogram/core.py:
```diff
@@ -32,6 +33,11 @@
 )
 
 
+def hist_safe_sum(items):
```
This just seems to drop some inputs?

martindurant commented on src/dask_histogram/core.py:
```diff
@@ -32,6 +33,11 @@
 
 
+def hist_safe_sum(items):
+    safe_items = [item for item in items if not isinstance(item, tuple)]
```
Directly summing the generator expr would be faster and more concise

martindurant commented on src/dask_histogram/core.py:
```diff
@@ -516,15 +560,11 @@ def histref(self):
     @property
     def _storage_type(self) -> type[bh.storage.Storage]:
         """Storage type of the histogram."""
-        if isinstance(self.histref, tuple):
```
I wonder why these existed in the first place

martindurant commented on src/dask_histogram/core.py:
```diff
@@ -878,6 +913,35 @@ def _partitionwise(func, layer_name, *args, **kwargs):
     )
 
 
+class PackedMultifill:
```
Isn't this also a lot like a partial?

lgray commented on src/dask_histogram/core.py:
```diff
@@ -330,9 +336,13 @@ def _blocked_dak_ma_w(
     )
-    return thehist.fill(*tuple(thedata), weight=theweights)
+
+    if ak.backend(*data) != "typetracer":
```
an empty fill is the same as returning a hist with no data in it.

lgray commented on src/dask_histogram/boost.py:
```diff
@@ -318,14 +327,14 @@ def fill(  # type: ignore
-        self._dask_name = new_name
+            self._staged.append(new_fill)
+        self._dask = None  # self._staged.__dask_graph__()
+        self._dask_name = "__not_yet_calculated__"
```
In terms of how we use histograms they really don't have any meaning until all fills we want to do are done.

They're stateful objects in terms of how we think about them, you fill a histogram and expect a histogram that has been filled to categorize and count what you fill it with. I think a functional approach would be confusing at best.

In terms of:
```
hist1 = ... # make something
hist2 = hist1 # some fork in the codepath
hist2.fill(..) # makes hist1 invalid
```
this only really happens in our workflows once a histogram has been filled (and in the language of dask also computed), and no further fills are done just manipulation on the counts and weights. Prior to that, the workflow is engaged just with filling the histograms.

lgray replied:
Now that adding isn't even allowed for dask-boost-histograms it makes this sort of clear.

lgray commented:
@martindurant Have now made it so that multifill is used for all the various input types... There are two red tests I need to figure out and then I think I've addressed your comments.

lgray commented on src/dask_histogram/boost.py:
```diff
@@ -242,12 +190,73 @@ def dask_name(self) -> str:
+
+            split_every = self._split_every
+            if split_every is None:
+                split_every = dask.config.get("histogram.aggregation.split-every", 8)
```
done!

lgray commented on src/dask_histogram/boost.py:
```diff
@@ -318,14 +327,14 @@ def fill(  # type: ignore
             raise ValueError(f"Cannot interpret input data: {args}")
 
-        new_fill = factory(*args, histref=self._histref, weights=weight, sample=sample)
+        # new_fill = partitioned_factory(*args, histref=self._histref, weights=weight, sample=sample)
```
done.

lgray commented:
Ran into a minor problem - couldn't add histograms any more if they have different partitions! The solution is to use the multi-source tree reduction I had cooked up before. This also helpfully gets rid of some of the new multi-fill routines.

lgray commented:
I am going to xfail those tests and add a corresponding error.

The functionality this is trying to cover is best done on the client side anyway.

martindurant commented on src/dask_histogram/boost.py:
```diff
@@ -112,11 +110,9 @@ def _histref(self):
-        elif not self.staged_fills() and other.staged_fills():
-            self._staged = other._staged
-        return self
+        raise NotImplementedError(
```
Could just remove add/iadd

lgray replied:
They are defined in the boost histogram base class, which would lead to some confusing behavior.

martindurant replied:
ok

martindurant commented on src/dask_histogram/core.py:
```diff
@@ -930,11 +984,20 @@ def _partitioned_histogram_multifill(
+
+        unpacked_multifill = partial(_blocked_multi_dak, repacker)
+        graph = dak_pwl(unpacked_multifill, name, *flattened_deps)
+    elif is_dataframe_like(flattened_deps[0]):
```
Do we have a test for this case?

lgray replied:
Yes - there are dataframe tests in `test_boost.py` that go through this code path.

martindurant reviewed (commented):
Not much to say, this time is was mostly moving things around

martindurant commented:
+1
